src/main/java/io/reactivex/rxjava3/internal/operators/observable/ObservableGroupBy.java

```diff
@@ -0,0 +1,391 @@
+    final int bufferSize;
+    final boolean delayError;
+
+    public ObservableGroupBy(ObservableSource<T> source,
+            Function<? super T, ? extends K> keySelector, Function<? super T, ? extends V> valueSelector,
+            int bufferSize, boolean delayError) {
+        super(source);
+        this.keySelector = keySelector;
+        this.valueSelector = valueSelector;
+        this.bufferSize = bufferSize;
+        this.delayError = delayError;
+    }
+
+    @Override
+    public void subscribeActual(Observer<? super GroupedObservable<K, V>> t) {
+        source.subscribe(new GroupByObserver<>(t, keySelector, valueSelector, bufferSize, delayError));
+    }
+
+    public static final class GroupByObserver<T, K, V> extends AtomicInteger implements Observer<T>, Disposable {
+
+        private static final long serialVersionUID = -3688291656102519502L;
+
+        final Observer<? super GroupedObservable<K, V>> downstream;
+        final Function<? super T, ? extends K> keySelector;
+        final Function<? super T, ? extends V> valueSelector;
+        final int bufferSize;
+        final boolean delayError;
+        final Map<Object, GroupedUnicast<K, V>> groups;
+
+        static final Object NULL_KEY = new Object();
+
+        Disposable upstream;
+
+        final AtomicBoolean cancelled = new AtomicBoolean();
+
+        public GroupByObserver(Observer<? super GroupedObservable<K, V>> actual, Function<? super T, ? extends K> keySelector, Function<? super T, ? extends V> valueSelector, int bufferSize, boolean delayError) {
+            this.downstream = actual;
+            this.keySelector = keySelector;
+            this.valueSelector = valueSelector;
+            this.bufferSize = bufferSize;
+            this.delayError = delayError;
+            this.groups = new ConcurrentHashMap<>();
+            this.lazySet(1);
+        }
+
+        @Override
+        public void onSubscribe(Disposable d) {
+            if (DisposableHelper.validate(this.upstream, d)) {
+                this.upstream = d;
+                downstream.onSubscribe(this);
+            }
+        }
+
+        @Override
+        public void onNext(T t) {
+            K key;
+            try {
+                key = keySelector.apply(t);
+            } catch (Throwable e) {
+                Exceptions.throwIfFatal(e);
+                upstream.dispose();
+                onError(e);
+                return;
+            }
+
+            Object mapKey = key != null ? key : NULL_KEY;
+            GroupedUnicast<K, V> group = groups.get(mapKey);
+            boolean newGroup = false;
+            if (group == null) {
+                // if the main has been cancelled, stop creating groups
+                // and skip this value
+                if (cancelled.get()) {
+                    return;
+                }
+
+                group = GroupedUnicast.createWith(key, bufferSize, this, delayError);
+                groups.put(mapKey, group);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **98, 100, and 108**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and uses that reference at line: **135**, **122**, **131**, and **128**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.